plugin/PlayAssetDelivery/src/main/java/com/google/play/core/godot/assetpacks/StateUpdateManager.java

```diff
@@ -0,0 +1,108 @@
+ *  	limitations under the License.
+ */
+
+/**
+ * This class keep track of all the ongoing asset pack requests and most updated asset pack states.
+ * Provides forceAssetPackStateUpdate() function to force-emit stateUpdated global signals, so that
+ * we are able to get the most updated asset pack state when the app is resumed from background.
+ */
+package com.google.play.core.godot.assetpacks;
+
+import static com.google.play.core.godot.assetpacks.PlayAssetDelivery.ASSET_PACK_STATE_UPDATED;
+
+import com.google.android.play.core.assetpacks.AssetPackManager;
+import com.google.android.play.core.assetpacks.AssetPackState;
+import com.google.android.play.core.assetpacks.model.AssetPackStatus;
+import com.google.play.core.godot.assetpacks.utils.PlayAssetDeliveryUtils;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import org.godotengine.godot.Dictionary;
+
+public class StateUpdateManager {
+  private Set<String> ongoingAssetPackRequests;
+  private Map<String, Dictionary> updatedAssetPackStateMap;
+  private final List<Integer> assetPackTerminalStates =
+      Arrays.asList(AssetPackStatus.COMPLETED, AssetPackStatus.FAILED, AssetPackStatus.CANCELED);
+  PlayAssetDelivery playAssetDeliveryPlugin;
+  private AssetPackManager assetPackManager;
+
+  public Set<String> ongoingAssetPackRequests() {
+    return ongoingAssetPackRequests;
+  }
+
+  public Map<String, Dictionary> updatedAssetPackStateMap() {
+    return updatedAssetPackStateMap;
+  }
+
+  public StateUpdateManager(
+      PlayAssetDelivery playAssetDeliveryPlugin, AssetPackManager assetPackManager) {
+    this.playAssetDeliveryPlugin = playAssetDeliveryPlugin;
+    this.assetPackManager = assetPackManager;
+    ongoingAssetPackRequests = Collections.synchronizedSet(new HashSet<>());
+    updatedAssetPackStateMap = new ConcurrentHashMap();
+  }
+
+  /** Package-private wrapper function used for argument captor. */
+  void emitSignalWrapper(String signalName, Object... signalArgs) {
+    playAssetDeliveryPlugin.emitSignalWrapper(signalName, signalArgs);
+  }
+  /**
+   * Function that emits assetPackStateUpdated signal if the given assetPackState has been updated.
+   */
+  public synchronized void emitNonDuplicateStateUpdatedSignal(
+      AssetPackState assetPackState, boolean addToOngoingAssetPackRequests) {
+    boolean isTerminalState = assetPackTerminalStates.contains(assetPackState.status());
+    if (isTerminalState) {
+      ongoingAssetPackRequests.remove(assetPackState);
+    }
+    if (!isTerminalState && addToOngoingAssetPackRequests) {
+      ongoingAssetPackRequests.add(assetPackState.name());
+    }
+    Dictionary assetPackStateDictionary =
+        PlayAssetDeliveryUtils.convertAssetPackStateToDictionary(assetPackState);
+    boolean isDifferentState =
```

Comment:
This boolean operation might be be clearer if you apply [DeMorgan's Law](https://en.wikipedia.org/wiki/De_Morgan%27s_laws). At least for me it's easier to read negations of individual booleans than negations of a combination of them:
````
boolean isDifferentState = !updatedAssetPackStateMap.containsKey(assetPackState.name())
                || updatedAssetPackStateMap.get(assetPackState.name()).hashCode() != assetPackStateDictionary.hashCode()
````

Reply:
I thought about using DeMorgan's when I first tried to refactor this boolean operation. However add that time I thought adding the negation in front of everything might be easier to read (maybe because I'm already familiar with the original boolean expression). 
Refactored this expression using DeMorgan's Law.